service/storage_service.cc

```diff
@@ -1339,7 +1339,7 @@ future<> storage_service::join_token_ring(sharded<db::system_distributed_keyspac
         if (_sys_ks.local().bootstrap_complete()) {
             throw std::runtime_error("Cannot replace address with a node that is already bootstrapped");
         }
```

Comment:
> However, a bootstrapping/replacing node has no persistent knowledge
of the down nodes, and it learns about their existance only from gossip.
But since the down nodes have unknown status, they never go
through `handle_state_normal` and therefore they are not accounted
as normal token owners.

Is this true for raft topology too?

I think for Raft topology the node should first become a member of Raft, and so can learn everything about the cluster, then start the replacement process.

Reply:
This was true also for raft topology at the time.
With all the recent changes there I'm not 100% sure that it's still the case.
Like I proposed elsewhere to @kbr-scylla, I think that we should move the endpoint_state map from gossiper to raft topology. With consistent-topology-changes enabled, raft topology, and only it, should manage the topology-related state while the gossiper can update non-topology related states.

Reply:
but endpoint_state map is a gossiper specific structure. It has stuff like generation or heartbeat or versioned values, none of which are relevant for raft-topology.

Reply:
> Is this true for raft topology too?

It's not. In Raft-topology, joining node (that includes node who wants to replace another node) is performing a handshake with the cluster by sending RPC to one of existing nodes, and it's the handler of the handshake who deals with ignore_nodes, verifies correctness of the request, liveness of dead nodes and so on.

Reply:
> but endpoint_state map is a gossiper specific structure. It has stuff like generation or heartbeat or versioned values, none of which are relevant for raft-topology.

Once raft topology is assuming ownership of the topology-related states (tokens, dc, rack) we can't say it's gossip-specific anymore.  The most major question is who is determining the node liveliness and its existence in the endpoint_state map.  With consistent topology, I'm certain that a. the responsibility should in one place, not dispersed between the gossiper and raft-topology, and b. it should move to raft-topology based on group0 membership.  Gossip can still be used in this model for states that are not critical for raft-topology, like heat map, compression dictionaries, etc.

Reply:
> Once raft topology is assuming ownership of the topology-related states (tokens, dc, rack) we can't say it's gossip-specific anymore.

Then you're not talking about endpoint_state map, but certain states that are distributed through gossip, and gossip internally uses endpoint_state map but that's an implementation detail of gossiper.

>  The most major question is who is determining the node liveliness and its existence in the endpoint_state map.

Node **liveness** is and will be determined by a failure detector, not by consensus. The consensus algorithm itself depends on failure detection to progress (leader election)